utilities/write_batch_with_index/write_batch_with_index.cc

```diff
@@ -52,28 +51,62 @@ class BaseDeltaIterator : public Iterator {
   }
 
   void SeekToFirst() override {
-    forward_ = true;
+    progress_ = Progress::SEEK_TO_FIRST;
     base_iterator_->SeekToFirst();
     delta_iterator_->SeekToFirst();
     UpdateCurrent();
   }
 
   void SeekToLast() override {
-    forward_ = false;
-    base_iterator_->SeekToLast();
-    delta_iterator_->SeekToLast();
+    progress_ = Progress::SEEK_TO_LAST;
+
+    // is there an upper bound constraint?
+    const Slice* base_upper_bound = base_iterator_upper_bound();
+    if (base_upper_bound != nullptr) {
+      // yes, and is base_iterator already constrained by an upper_bound?
+      if (!base_iterator_->check_upper_bound()) {
+        // no, so we have to seek it to before base_upper_bound
+        base_iterator_->Seek(*(base_upper_bound));
+        if (base_iterator_->Valid()) {
+          base_iterator_->Prev();  // upper bound should be exclusive!
+        } else {
+          // the base_upper_bound is beyond the base_iterator, so just
+          // SeekToLast()
+          base_iterator_->SeekToLast();
+        }
+      } else {
+        // yes, so the base_iterator will take care of base_upper_bound
+        base_iterator_->SeekToLast();
+      }
+
+      // delta iterator does not itself support iterate_upper_bound,
+      // so we have to seek it to before iterate_upper_bound
+      delta_iterator_->Seek(*(read_options_->iterate_upper_bound));
```

Comment:
I think this value could be nullptr, if the base iterator had an upper bound but the read_options did not.

Reply:
I have fixed that one now too